Spigot-API-Patches/0128-Allow-Blocks-to-be-accessed-via-a-long-key.patch

```diff
@@ -0,0 +1,116 @@
+From 70ad8fff4d941055e069871a70ee4518185170d5 Mon Sep 17 00:00:00 2001
+From: Spottedleaf <Spottedleaf@users.noreply.github.com>
+Date: Tue, 14 Aug 2018 21:42:10 -0700
+Subject: [PATCH] Allow Blocks to be accessed via a long key
+
+The key can be retrieved via methods Location#toBlockKey() and
+Block#getBlockKey()
+
+World provides lookup for blocks by long key via method World#getBlockAtKey(long)
+
+The formatting for the key is as follows:
+
+4 bits unused|8 bit y|26 bit z|26 bit x
+
+The y value is considered unsigned while z and x are considered two's complement
+
+Y range: [0, 255]
+X, Z range: [-33 554 432, 33 554 431]
+
+diff --git a/src/main/java/org/bukkit/Location.java b/src/main/java/org/bukkit/Location.java
+index 253f0c2..4deb4f7 100644
+--- a/src/main/java/org/bukkit/Location.java
++++ b/src/main/java/org/bukkit/Location.java
+@@ -10,7 +10,6 @@ import org.bukkit.util.Vector;
+ 
+ // Paper start
+ import java.util.Collection;
+-import java.util.Collections;
+ import java.util.function.Predicate;
+ import org.bukkit.entity.Entity;
+ import org.bukkit.entity.LivingEntity;
+@@ -558,6 +557,18 @@ public class Location implements Cloneable, ConfigurationSerializable {
+         blockLoc.setZ(getBlockZ());
+         return blockLoc;
+     }
++
++    // Paper Start
++
++    /**
++     * @return The block key for this location's block location.
++     * @see Block#getBlockKey()
++     */
++    public long toBlockKey() {
++        return ((long)getBlockX() & 0x3FFFFFFL) | (((long)getBlockY() & 0xFFL) << 52) | (((long)getBlockZ() & 0x3FFFFFFL) << 26);
```

Comment:
I hate that you ordered it this way. I think having the two big ones first makes more sense.